src/dispatch/batch_builder.rs

```diff
@@ -0,0 +1,496 @@
+    fn writes() -> Vec<ResourceId> {
+        vec![]
+    }
+}
+
+/// The `BatchSystem` allow to register some subsystems that it will take
+/// care to execute once the main dispatcher execute it.
+/// It has the capability to restart its systems many times, and this
+/// can be controlled by the user.
+///
+/// The BatchSystem keep track of its internal dependecies and is able to run
+/// in parallel with other systems.
+/// At the same time the subsystems can run in parallel within the BatchSystem.
+///
+/// To add a new batch system you have to call the function `with_batch` which
+/// accept a `BatchBuilder` object that will take care to construct the `BatchSystem`.
+pub struct BatchSystem<'a> {
+    stages: Vec<Stage<'a>>,
+    batch_controller: Box<for<'c> BatchControllerSystem<'c> + Send + 'a>,
+    running_time: RunningTime,
+
+    #[cfg(feature = "parallel")]
+    thread_pool: ::std::sync::Arc<::rayon::ThreadPool>,
+}
+
+impl<'a> BatchSystem<'a> {
+    #[cfg(feature = "parallel")]
+    pub(crate) fn new(
+        stages: Vec<Stage<'a>>,
+        batch_controller: Box<for<'c> BatchControllerSystem<'c> + Send + 'a>,
+        running_time: RunningTime,
+        thread_pool: ::std::sync::Arc<::rayon::ThreadPool>,
+    ) -> Self {
+        BatchSystem {
+            stages,
+            batch_controller,
+            running_time,
+            thread_pool,
+        }
+    }
+
+    #[cfg(not(feature = "parallel"))]
+    pub(crate) fn new(
+        stages: Vec<Stage<'a>>,
+        batch_controller: Box<for<'c> BatchControllerSystem<'c> + Send + 'a>,
+        running_time: RunningTime,
+    ) -> Self {
+        Batch {
+            stages,
+            batch_controller,
+            running_time,
+        }
+    }
+
+    /// Dispatches the systems (except thread local systems)
+    /// in parallel given the resources to operate on.
+    ///
+    /// This operation blocks the
+    /// executing thread.
+    ///
+    /// Only available with "parallel" feature enabled.
+    ///
+    /// Please note that this method assumes that no resource
+    /// is currently borrowed. If that's the case, it panics.
+    #[cfg(feature = "parallel")]
+    pub fn dispatch_par(&mut self, world: &World) {
+        let stages = &mut self.stages;
+
+        self.thread_pool.install(move || {
+            for stage in stages {
+                stage.execute(world);
+            }
+        });
+    }
+
+    /// Dispatches the systems (except thread local systems) sequentially.
+    ///
+    /// This is useful if parallel overhead is
+    /// too big or the platform does not support multithreading.
+    ///
+    /// Please note that this method assumes that no resource
+    /// is currently borrowed. If that's the case, it panics.
+    pub fn dispatch_seq(&mut self, world: &World) {
+        for stage in &mut self.stages {
+            stage.execute_seq(world);
+        }
+    }
+}
+
+impl<'a> System<'a> for BatchSystem<'_> {
+    type SystemData = UncheckedWorld<'a>;
```

Comment:
This doesn't quite work, because now the `BatchSystem` has no resource requirements, thus can run in parallel with everything else (which it shouldn't).

You need to collect all the `reads` and `writes` of all inner system, sort them and remove all the duplicates. Then, you can return those resource ids using a

* custom [`Accessor`](https://docs.rs/shred/0.9.2/shred/trait.Accessor.html) implementation (a struct with two `Vec`s, which you just return from `Accessor::{reads, writes}`)
* custom [`DynamicSystemData`](https://docs.rs/shred/0.9.2/shred/trait.DynamicSystemData.html), which just defines the associated `Accessor` and sets it to your custom implementation